File: 773/Solution.java
class Solution {
    public int slidingPuzzle(int[][] board) {
        Queue<String> q = new LinkedList<>();
        Set<String> set = new HashSet<>();
        String s = "";
        String target = "123450";
        int[][] dir = {{1,3}, {0,2,4}, {1,5}, {0,4}, {1,3,5}, {2,4}};
        for(int[] a : board) for(int i : a) s += i;
        set.add(s);
        q.add(s);
        int step = 0;
        while(!q.isEmpty()){
            int size = q.size();
            for(int i = 0; i < size; i++){
                String cur = q.poll();
                if(cur.equals(target)) return step;
                int index = cur.indexOf('0');
                for(int move: dir[index]){
                    StringBuilder next = new StringBuilder(cur);
                    next.setCharAt(index, next.charAt(move));
                    next.setCharAt(move, '0');

                    if(!set.contains(next.toString())){
                        set.add(next.toString());
                        q.offer(next.toString());
                    }
                }
            }   
            step++;
        }
        return -1;

    }
}

File: 773/Solution.go
func slidingPuzzle(board [][]int) int {
    hm := make(map[string]bool)
    q := make([]string, 0)
    s , target := "", "123450"
    for _, r := range board {
        for _, val := range r {
            s += strconv.Itoa(val)
        }
    }

    hm[s] = true
    q = append(q, s)

    step := 0
    dir := [6][]int{{1,3}, {0,2,4}, {1,5}, {0,4}, {1,3,5}, {2,4}}

    for len(q) > 0 {
        size := len(q)
        for size > 0 {
            cur := q[0]
            q = q[1:]
            if cur == target {
                return step
            }
            zeroInd := strings.IndexRune(cur, '0')
            for _, i:= range dir[zeroInd] {
                ss:= []byte(cur)
                ss[zeroInd], ss[i] = ss[i], ss[zeroInd]
                next := string(ss)
                if !hm[next] {
                    hm[next] = true
                    q = append(q, next)
                }
            }
            size--
        }
        step++
    }
    return -1
}
